### ISSUE TITLE
[BUG] Chromium and webkit headless/head rendering differences

### ISSUE THREAD

**Issue by manuman94** [microsoft/playwright#12683]
## System:
 - OS: Linux 5.4 Ubuntu 20.04.3 LTS (Focal Fossa)
 - Memory: 1.12 GB / 15.45 GB

## Binaries:
 - Node: 14.18.0 - /usr/local/bin/node
 - Yarn: 1.22.5 - /usr/bin/yarn
 - npm: 8.5.1 - /usr/local/bin/npm
## Languages:
 - Bash: 5.0.17 - /bin/bash

## Playwright env.:
- Playwright Version: 1.19.2
- Browser: All

## The bug

Hi there! We are testing a kind of "analytics" web where some charts and diagrams are shown using amCharts library. The first approach is being made with snapshot testing. 

The **issue** comes with **head/headless option**, as we develop with headless option set to `false` and the CI environment uses headless executions. 

**When the snapshots are made with headless option enabled and then we launch the tests in headful mode the tests in Chromium and Webkit are failing, the images are not equal**. This doesn't happen with Firefox, that makes identically component screenshots.

## Evidences

We made a simple project that enters an amChart example and makes a snapshot of a pie chart component. The test consist on comparing a snapshot made with headless option set to `true` and then repeat the test in headful mode (and vice versa).

https://github.com/manuman94/playwright-head-vs-headless-rendering

Further, I tried to compare the images using pixelmatch library (which Playwright is using too I believe) and get the image difference. Here you can see the both images and the difference image:
Headless:
![headless](https://user-images.githubusercontent.com/6721904/157857322-aae6224e-54bf-4eb3-96f9-a244d9edf7c3.jpeg)

HeadFul:
![noHeadless](https://user-images.githubusercontent.com/6721904/157857310-945baa23-ec08-4ed9-b1da-3659957bdcec.jpeg)

Difference:
![difference](https://user-images.githubusercontent.com/6721904/157857333-dcb4d673-3856-43c7-8b00-900a9f0609b7.jpeg)

## Previous issues
There was an issue opened on Jan 4th 2021 (https://github.com/microsoft/playwright/issues/11177) regarding this issue, but as for today is closed and a pull request was created to solve the font family problems.

There is another closed issue (https://github.com/microsoft/playwright/issues/5371) but it seems that it was reporting a completely different problem.

## Side notes
- Consequent screenshots in the same mode (headless or headful) are identical. This confirms that the amCharts library renders its components in the same way every time.
- The font family is not the only issue as the difference image shows that some pixels part of the chart are different too.
- Firefox does not have this difference.
- Is there any way of obtaining the result image using the matcher .toMatchSnapshot? The method signature tells that it returns a Buffer but I'm not able to obtain the difference image. The one I got is using pixelmatch function directly.

Thank you so much, greetings.

**Comment by yury-s** [comment]
Do you see the screenshots that don't include `<canvas/>` element?

Looks like another issue caused by different anti-aliasing behavior is different between headed and headless modes.  Unfortunately there is no good way to make the browsers produce identical images in both modes. We are adding additional settings to `toMatchSnapshot` matcher in the next release (see #12167), namely `pixelCount` and `pixelRatio`. They will enable you to increase the tolerance level of the diff algorithm to allow for small discrepancies like this. You can already give it a try with `npm i @playwright/test@next`

**Comment by manuman94** [comment]
Hi @yury-s,

Yes, we have a test that uses SVGs and the screenshots doesn't match between headful and headless. In fact, the screenshots above are of an SVG, it's not a canvas object.

Short HTML code behind the screenshots: 
![image](https://user-images.githubusercontent.com/6721904/158128354-13bc9c1f-c675-43dd-9c73-4b6bfb19b046.png)

Okay! So the only way to avoid this issue is to play with tolerance, as I understand it is a browser issue.

Thank you so much for your support. 

Greetings.

**Comment by aslushnikov** [comment]
@manuman94 the `maxDiffPixels` and `maxDiffPixelRatio` were released yesterday in 1.20 release.

**Comment by aslushnikov** [comment]
Firefox screenshots do differ as well on headless vs headed under certain circumstances. 
- [ ] Consider including headlessness to the upcoming `screenshotsDir` path

**Comment by amrsa1** [comment]
@aslushnikov 
Seems that maxDiffPixels doesn't have any effect on the test 

<img width="1398" alt="Screenshot 2022-05-03 at 16 23 11" src="https://user-images.githubusercontent.com/32042806/166472393-318a2c84-3d37-498a-ba94-6fc9213935ab.png">

the 15 pixels are the difference between headed and headless mode

**Comment by aslushnikov** [comment]
> the 15 pixels are the difference between headed and headless mode

`maxDiffPixels` only comes into play when image dimensions are equal.

**Comment by benwiley4000** [comment]
The problem with our case is that headed vs headless renders the image with one pixel difference of dimensions. I'm not sure I understand how `maxDiffPixelRatio` will work. Does it also require `maxDiffPixels` to be set? Or does it replace this?

**Comment by cassianomon** [comment]
I am also having the same issue, the main problem is one of the modals show a scrollbar (which is the defect I want to capture), it gets caught when I run with headed mode on Chromium and Firefox, but it doesn't caught the issue when it's headless (when it's running on CI for example). Having a option to enable scrollbars on headless mode would be good, as we're testing Visual here, it makes total sense.

**Comment by gselsidi** [comment]
what is the status on this? Noticing these issues when creating baseline images from headless / headed and vice versa. Normally I'm able to get away with an insane .0001 maxpixeldifference!!

When running cross headed / headless not even .01 get's them to pass

**Comment by gselsidi** [comment]
I'm noticing maxDiffPixels: 5, is not taking.... On ci there is a 1 pixel difference of the snapshot say 1x1 vs 1x2 and it fails.

Expected an image 60px by 125px, received 60px by 124px......

**Comment by IvanAbramow** [comment]
@aslushnikov 
Hello! I have exactly the same problem, the size of the element differs by 1 - 2 pixels.
Screenshot comparison failed:  Expected an image 640px by 694px, received 640px by 696px.

This appears for both page.screenshot and element.screenshot. 

For example, in jest-image-snapshot package https://www.npmjs.com/package/jest-image-snapshot
there is an **allowSizeMismatch** parameter that allows you to solve this problem.

@playwright/test 1.27.1

Please find a solution to fix this problem as soon as possible. Due to this problem, I can not completely switch to playwright without using jest

**Comment by dgtlmoon** [comment]
> Please find a solution to fix this problem as soon as possible.

I think this is an inappropriate comment considering that the software is provided to you free of charge and without warranty

**Comment by IvanAbramow** [comment]
> > Please find a solution to fix this problem as soon as possible.
> 
> I think this is an inappropriate comment considering that the software is provided to you free of charge and without warranty

I agree with you, I don't mean to offend the playwright team. My mistake. I really love this product and enjoy using it. I just found a few requests about this problem, so I asked to try to solve it.

**Comment by eveltman-twinfield** [comment]
Would the new Chrome headless mode (not yet officially supported in Playwright) solve this issue?
https://developer.chrome.com/articles/new-headless/#whats-new-in-headless

As I understand it, the performance drops quite a bit, but for me, if the rendering becomes identical to the headful browser, and if it increases the usability of visual testing (developing in headful mode and running in CI in headless mode), then I would prefer the new headless mode of Chrome.

**Comment by JonMarbach** [comment]
[aslushnikov](https://github.com/aslushnikov) +1 to "shall we introduce an option to enable scrollbars in headless mode?"

But perhaps more importantly, the new headless mode that @eveltman-twinfield mentioned above sounds even better. A relevant quote from that article: "This created a confusing situation where any automated browser test might pass in headful mode but fail in Headless mode, or vice versa — a major pain point for automation engineers."

**Comment by rgustavsson** [comment]
What's needed to try out the new `chrome --headless=new` flag?
@eveltman-twinfield mentioned that this isn't officially supported in Playwright yet, but should I be able to try it by using the latest Playwright version, setting browser launch option `Headless` to false, and then pass `--headless=new` in the `Args` browser launch option? :)

**Comment by chartinger** [comment]
Since https://github.com/microsoft/playwright/commit/b21c81d3448dd47277997683d042c21aab095978 there is a new env var option `PLAYWRIGHT_CHROMIUM_USE_HEADLESS_NEW` but it seems to be undocumented (?)

**Comment by afonte15** [comment]
Hi, are there any updates on this bug.  I am currently running into this issue between headless and headed modes. I am currently experiencing this issue.

**Comment by oliverp** [comment]
Hello, im running in this issue with the width difference in headless creation & headless test. So taking snapshots with a headless environment on the same machine and testing against those snapshots on the same machine produces :
``` 
Expected an image 1125px by 229px, received 1133px by 229px.
``` 
I tested to define the viewport without success.
I tested chrome without success.

I guess I will crop the resulting snapshots, so the test might work.

**Comment by barbolo** [comment]
> Since [b21c81d](https://github.com/microsoft/playwright/commit/b21c81d3448dd47277997683d042c21aab095978) there is a new env var option `PLAYWRIGHT_CHROMIUM_USE_HEADLESS_NEW` but it seems to be undocumented (?)

https://developer.chrome.com/articles/new-headless/

**Comment by gselsidi** [comment]
would this solve anything about cross architecture differences? Because in the past even CPU differences for the same architecture would cause slight differences and it would fail. 

So if i author on say Mac and then run on Linux in CI would this new chrome feature resolve those issues? Has anyone tried it?
